extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java

```diff
@@ -1,189 +1,266 @@
-          KuduPredicate.newComparisonPredicate(
-              schema.getColumn(adapterIdField),
-              ComparisonOp.EQUAL,
-              adapterId);
-      if (!isDataIndex && ranges != null && !ranges.isEmpty()) {
-        for (final SinglePartitionQueryRanges r : ranges) {
-          final byte[] partitionKey =
-              ((r.getPartitionKey() == null) || (r.getPartitionKey().length == 0))
-                  ? KuduUtils.EMPTY_KEY
-                  : r.getPartitionKey();
-          for (final ByteArrayRange range : r.getSortKeyRanges()) {
-            final byte[] start = range.getStart() != null ? range.getStart() : new byte[0];
-            final byte[] end =
-                range.getEnd() != null ? range.getEndAsNextPrefix()
-                    : new byte[] {
-                        (byte) 0xFF,
-                        (byte) 0xFF,
-                        (byte) 0xFF,
-                        (byte) 0xFF,
-                        (byte) 0xFF,
-                        (byte) 0xFF,
-                        (byte) 0xFF};
-            KuduPredicate lowerPred =
-                KuduPredicate.newComparisonPredicate(
-                    schema.getColumn(KuduField.GW_SORT_KEY.getFieldName()),
-                    ComparisonOp.GREATER_EQUAL,
-                    start);
-            KuduPredicate upperPred =
-                KuduPredicate.newComparisonPredicate(
-                    schema.getColumn(KuduField.GW_SORT_KEY.getFieldName()),
-                    ComparisonOp.LESS,
-                    end);
-            KuduPredicate partitionPred =
-                KuduPredicate.newComparisonPredicate(
-                    schema.getColumn(KuduField.GW_PARTITION_ID_KEY.getFieldName()),
-                    ComparisonOp.EQUAL,
-                    partitionKey);
-
-            KuduScannerBuilder scannerBuilder = operations.getScannerBuilder(table);
-            KuduScanner scanner =
-                scannerBuilder.addPredicate(adapterIdPred).addPredicate(lowerPred).addPredicate(
-                    upperPred).addPredicate(partitionPred).build();
-            KuduUtils.executeQuery(scanner, results);
-          }
-        }
-      } else if (dataIds != null) {
-        final String partitionIdField =
-            isDataIndex ? KuduDataIndexField.GW_PARTITION_ID_KEY.getFieldName()
-                : KuduField.GW_PARTITION_ID_KEY.getFieldName();
-        for (final byte[] dataId : dataIds) {
-          KuduPredicate partitionPred =
+    final List<AsyncKuduScanner> scanners = new ArrayList<>();
+
+    KuduPredicate adapterIdPred =
+        KuduPredicate.newInListPredicate(
+            schema.getColumn(KuduField.GW_ADAPTER_ID_KEY.getFieldName()),
+            Arrays.asList(ArrayUtils.toObject(adapterIds)));
+    if ((ranges != null) && !ranges.isEmpty()) {
+      for (final SinglePartitionQueryRanges r : ranges) {
+        final byte[] partitionKey =
+            ((r.getPartitionKey() == null) || (r.getPartitionKey().length == 0))
```

Comment:
Let's pull out `r.getPartitionKey()` into a variable, so we don't have to call it multiple times.

Reply:
I think we can even remove the check for `r.getPartitionKey().length == 0` since that is implicitly an `EMPTY_KEY`